pkg/dogstatsd/filter/filter.go

```diff
@@ -0,0 +1,62 @@
+package filter
+
+import (
+	"regexp"
+	"strings"
+)
+
+type TagFilter struct {
+	filters []*regexp.Regexp
+}
+
+func NewTagFilter(regexes []string) (*TagFilter, error) {
+	filters := make([]*regexp.Regexp, len(regexes))
+	for i, r := range regexes {
+		regex, err := regexp.Compile(r)
+		if err != nil {
+			return nil, err
+		}
+
+		filters[i] = regex
+	}
+	return &TagFilter{
+		filters: filters,
+	}, nil
+}
+
+func (tf *TagFilter) Filter(tags []string) []string {
+	var i int
+	for _, t := range tags {
+		if ok, result := tf.filter(t); ok {
+			tags[i] = result
+			i++
+		}
+	}
+
+	// Prevent memory leak by erasing truncated values
+	for j := i; j < len(tags); j++ {
+		tags[j] = ""
+	}
+
+	return tags[:i]
+}
+
+func (tf *TagFilter) filter(tag string) (bool, string) {
+OUTER:
+	for _, regex := range tf.filters {
+		if m := regex.FindStringSubmatch(tag); m != nil {
+			for i, name := range regex.SubexpNames() {
+				if name == "Keep" {
+					if strings.Contains(m[i], ":") {
+						tag = m[i]
+					}
+					continue OUTER
```

Comment:
question: Since matches without a `Keep` return immediately, why not apply the same behavior to `Keep` matches and drop the looping complexity a bit? Is it an intentional feature to have a regex extract a value from a tag only to then match a dropping rule?